graphql_compiler/tests/test_ir_generation.py

```diff
@@ -2251,3 +2251,757 @@ def test_coercion_on_union_within_fold_scope(self):
+                    expressions.NullLiteral
+                ),
+            }),
+        ]
+        expected_location_types = {
+            base_location: 'Animal',
+            parent_location: 'Animal',
+            sibling_location: 'Animal',
+            sibling_species: 'Species',
+            revisited_base_location: 'Animal'
+        }
+
+        check_test_data(self, test_data, expected_blocks, expected_location_types)
+
+    def test_traverse_and_optional_and_traverse(self):
+        test_data = test_input_data.traverse_and_optional_and_traverse()
+
+        base_location = helpers.Location(('Animal',))
+        parent_location = base_location.navigate_to_subpath('in_Animal_ParentOf')
+        sibling_location = parent_location.navigate_to_subpath('out_Animal_ParentOf')
+        sibling_species = sibling_location.navigate_to_subpath('out_Animal_OfSpecies')
+        revisited_parent_location = parent_location.revisit()
+
+        expected_blocks = [
+            blocks.QueryRoot({'Animal'}),
+            blocks.MarkLocation(base_location),
+            blocks.Traverse('in', 'Animal_ParentOf'),
+            blocks.MarkLocation(parent_location),
+            blocks.Traverse('out', 'Animal_ParentOf', True),
+            blocks.MarkLocation(sibling_location),
+            blocks.Traverse('out', 'Animal_OfSpecies'),
+            blocks.MarkLocation(sibling_species),
+            blocks.Backtrack(sibling_location),
+            blocks.EndOptional(),
+            blocks.Backtrack(parent_location, True),
+            blocks.MarkLocation(revisited_parent_location),
+            blocks.Backtrack(base_location),
+            blocks.ConstructResult({
+                'sibling_name': expressions.TernaryConditional(
+                    expressions.ContextFieldExistence(sibling_location),
+                    expressions.OutputContextField(
+                        sibling_location.navigate_to_field('name'), GraphQLString),
+                    expressions.NullLiteral
+                ),
+                'animal_name': expressions.OutputContextField(
+                        base_location.navigate_to_field('name'), GraphQLString
+                ),
+                'sibling_and_self_species': expressions.TernaryConditional(
+                    expressions.ContextFieldExistence(sibling_species),
+                    expressions.OutputContextField(
+                        sibling_species.navigate_to_field('name'), GraphQLString),
+                    expressions.NullLiteral
+                ),
+                'parent_name': expressions.OutputContextField(
+                    parent_location.navigate_to_field('name'), GraphQLString
+                ),
+            }),
+        ]
+        expected_location_types = {
+            base_location: 'Animal',
+            parent_location: 'Animal',
+            sibling_location: 'Animal',
+            sibling_species: 'Species',
+            revisited_parent_location: 'Animal'
```

Comment:
nit: it's nice to include a trailing `,` even for the last element -- that way, if someone wants to add another element on the line below, they don't have to also change this line, and `git blame` points to the actual author of the line instead of the person who added a `,`